File: android-app/feature/feature-discover/src/main/java/tv/bayit/plus/feature/discover/data/FeatureAvailabilityService.kt
package tv.bayit.plus.feature.discover.data

import tv.bayit.plus.feature.discover.model.DiscoverFeature
import tv.bayit.plus.feature.discover.model.DiscoverPlatform
import tv.bayit.plus.feature.discover.model.FeatureAvailabilityState
import tv.bayit.plus.feature.discover.model.FeaturePrerequisite
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Determines whether a [DiscoverFeature] is currently accessible to the user.
 *
 * Evaluation proceeds in this order, mirroring the iOS `FeatureAvailabilityService`:
 *
 * 1. Platform check — if [DiscoverPlatform.ANDROID] is not in [DiscoverFeature.platforms],
 *    returns [FeatureAvailabilityState.PlatformOnly] immediately.
 * 2. Prerequisites — iterates [DiscoverFeature.prerequisites] and:
 *    - A missing [FeaturePrerequisite.PrerequisiteType.SUBSCRIPTION] returns
 *      [FeatureAvailabilityState.PremiumRequired] immediately (hard gate).
 *    - A [FeaturePrerequisite.PrerequisiteType.CONTENT_TYPE] prerequisite returns
 *      [FeatureAvailabilityState.NotAvailable] immediately, because context-dependent
 *      features cannot be unlocked through setup actions.
 *    - All other unmet prerequisites are accumulated into the [missing] list.
 * 3. If [missing] is non-empty, returns [FeatureAvailabilityState.SetupNeeded].
 * 4. Otherwise returns [FeatureAvailabilityState.Ready].
 */
@Singleton
class FeatureAvailabilityService @Inject constructor(
    private val dependencies: AvailabilityDependencies,
) {

    /**
     * Evaluates [feature] against the current user state.
     *
     * Suspend functions in [dependencies] may make network requests; callers
     * should invoke this from a coroutine scope with an appropriate dispatcher.
     */
    suspend fun checkAvailability(feature: DiscoverFeature): FeatureAvailabilityState {
        if (DiscoverPlatform.ANDROID !in feature.platforms) {
            return platformOnlyState(feature.platforms)
        }

        val missing = mutableListOf<FeaturePrerequisite>()

        for (prerequisite in feature.prerequisites) {
            when (prerequisite.type) {
                FeaturePrerequisite.PrerequisiteType.SUBSCRIPTION -> {
                    if (!dependencies.isPremium()) {
                        return FeatureAvailabilityState.PremiumRequired
                    }
                }

                FeaturePrerequisite.PrerequisiteType.CONTENT_TYPE -> {
                    return FeatureAvailabilityState.NotAvailable(
                        reasonKey = prerequisite.labelKey,
                    )
                }

                FeaturePrerequisite.PrerequisiteType.AVATAR -> {
                    if (!dependencies.hasAvatar()) {
                        missing += prerequisite
                    }
                }

                FeaturePrerequisite.PrerequisiteType.MICROPHONE -> {
                    if (!dependencies.hasMicrophonePermission()) {
                        missing += prerequisite
                    }
                }

                FeaturePrerequisite.PrerequisiteType.PREFERENCE -> {
                    if (!dependencies.hasCompletedPreference(prerequisite.id)) {
                        missing += prerequisite
                    }
                }

                FeaturePrerequisite.PrerequisiteType.VOICE_CLONE -> {
                    if (!dependencies.hasAvatar()) {
                        missing += prerequisite
                    }
                }
            }
        }

        return if (missing.isNotEmpty()) {
            FeatureAvailabilityState.SetupNeeded(missing)
        } else {
            FeatureAvailabilityState.Ready
        }
    }

    private fun platformOnlyState(platforms: Set<DiscoverPlatform>): FeatureAvailabilityState {
        val representative = platforms.firstOrNull()
            ?: return FeatureAvailabilityState.NotAvailable(
                reasonKey = "discover.availability.noPlatform",
            )
        return FeatureAvailabilityState.PlatformOnly(representative)
    }
}

File: ios-app/Packages/BayitCore/Sources/BayitCore/Discover/FeatureAvailabilityService.swift
import Foundation

public protocol FeatureAvailabilityChecking: Sendable {
    func checkAvailability(for feature: DiscoverFeature) async -> FeatureAvailabilityState
}

public struct FeatureAvailabilityDependencies: Sendable {
    public let isPremium: @Sendable () async -> Bool
    public let hasAvatar: @Sendable () async -> Bool
    public let hasMicrophonePermission: @Sendable () async -> Bool
    public let hasCompletedPreference: @Sendable (String) async -> Bool

    public init(
        isPremium: @escaping @Sendable () async -> Bool,
        hasAvatar: @escaping @Sendable () async -> Bool,
        hasMicrophonePermission: @escaping @Sendable () async -> Bool,
        hasCompletedPreference: @escaping @Sendable (String) async -> Bool
    ) {
        self.isPremium = isPremium
        self.hasAvatar = hasAvatar
        self.hasMicrophonePermission = hasMicrophonePermission
        self.hasCompletedPreference = hasCompletedPreference
    }
}

public final class FeatureAvailabilityService: FeatureAvailabilityChecking, Sendable {
    #if os(tvOS)
        private static let currentPlatform: Platform = .tvOS
    #else
        private static let currentPlatform: Platform = .iOS
    #endif

    private let dependencies: FeatureAvailabilityDependencies

    public init(dependencies: FeatureAvailabilityDependencies) {
        self.dependencies = dependencies
    }

    public func checkAvailability(
        for feature: DiscoverFeature
    ) async -> FeatureAvailabilityState {
        guard feature.platforms.contains(Self.currentPlatform) else {
            return platformRequirement(from: feature.platforms)
        }

        var missingPrerequisites: [FeaturePrerequisite] = []

        for prerequisite in feature.prerequisites {
            switch prerequisite.type {
            case .subscription:
                let premium = await dependencies.isPremium()
                if !premium {
                    return .premiumRequired
                }

            case .contentType:
                return .notAvailable(
                    reasonKey: prerequisite.labelKey
                )

            case .avatar:
                let avatarReady = await dependencies.hasAvatar()
                if !avatarReady {
                    missingPrerequisites.append(prerequisite)
                }

            case .microphone:
                let micReady = await dependencies.hasMicrophonePermission()
                if !micReady {
                    missingPrerequisites.append(prerequisite)
                }

            case .preference:
                let completed = await dependencies.hasCompletedPreference(
                    prerequisite.id
                )
                if !completed {
                    missingPrerequisites.append(prerequisite)
                }

            case .voiceClone:
                let avatarReady = await dependencies.hasAvatar()
                if !avatarReady {
                    missingPrerequisites.append(prerequisite)
                }
            }
        }

        if !missingPrerequisites.isEmpty {
            return .setupNeeded(missingPrerequisites)
        }

        return .ready
    }

    private func platformRequirement(
        from platforms: Set<Platform>
    ) -> FeatureAvailabilityState {
        if let required = platforms.first {
            return .platformOnly(required)
        }
        return .notAvailable(reasonKey: "discover.availability.noPlatform")
    }
}
